util/sqlexec/restricted_sql_executor.go

```diff
@@ -35,6 +35,10 @@ import (
 type RestrictedSQLExecutor interface {
 	// ExecRestrictedSQL run sql statement in ctx with some restriction.
 	ExecRestrictedSQL(ctx sessionctx.Context, sql string) ([]chunk.Row, []*ast.ResultField, error)
+	// ExecRestrictedSQLWithSnapshot run sql statement in ctx with some restriction and with snapshot.
+	// If current session sets the snapshot timestamp, then execute with this snapshot timestamp.
+	// Otherwise, execute with the current transaction start timestamp if the transaction is valid.
+	ExecRestrictedSQLWithSnapshot(ctx sessionctx.Context, sql string) ([]chunk.Row, []*ast.ResultField, error)
```

Comment:
Why adding a new method to the interface, rather than 
```
ExecRestrictedSQL("set @@tidb_snapshot = now(); sql")
```

Reply:
What if the sql executed failed? Then the session's snapthotTS will not be clear, then put the session back to session pool. Next time get the session from pool to execute some sql may return error result.

Reply:
So how about put the `set tidb_snapshot` outside to the caller and remove this method?
We can make the session pool use the context from `ctx sessionctx.Context` argument